torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -0,0 +1,922 @@
+# Owner(s): ["module: onnx"]
+from __future__ import annotations
+
+import dataclasses
+import inspect
+import logging
+from types import ModuleType
+
+from typing import Callable, Mapping, Optional, Sequence, Set
+
+import torch
+import torch._ops
+import torch.fx
+
+from torch import _prims_common, _refs
+
+from torch._prims_common import (
+    ELEMENTWISE_TYPE_PROMOTION_KIND,
+    wrappers as _prims_common_wrappers,
+)
+from torch._refs import linalg as _linalg_refs, nn as _nn_refs, special as _special_refs
+from torch._refs.nn import functional as _functional_refs
+
+# Imported to resolve beartype issue when type checking node.Argument.
+from torch.fx.node import Node  # noqa: F401
+
+from torch.onnx._internal import _beartype
+from torch.utils import _pytree
+
+logger = logging.getLogger(__name__)
+
+
+def _try_getclosurevars(func):
+    try:
+        return inspect.getclosurevars(func)
+    except TypeError as e:
+        return None
+
+
+@dataclasses.dataclass
+class TypePromotionSnapshot:
+    """Type promotion snapshot for a fx node and its inputs.
+
+    Contains the promoted dtype for args and kwargs that needs promoting.
+    Contains the expected node output dtype.
+    """
+
+    args_dtypes: Mapping[int, torch.dtype]
+    """Mapping from arg position to dtype to promote to."""
+
+    kwargs_dtypes: Mapping[str, torch.dtype]
+    """Mapping from kwarg name to dtype to promote to."""
+
+    out_dtype: torch.dtype
+    """Expected output dtype of the node."""
+
+
+class TypePromotionRule:
+    """Defines how to perform type promotion for 'torch.ops.{namespace}.{op_name}'."""
```

Comment:
Maybe document how to "interpret" the promotion rules`so that we others can easily check whether the rules are correct, etc

```python
    TypePromotionRule(
        "aten", "bitwise_and_", [0, 1], [], ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
    ),
    TypePromotionRule(
        "aten",
        "bitwise_left_shift",
        [0, 1],
        [],
        ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
    ),
```

Reply:
Also document how `promotion_kind` works and its different values such as `ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT, ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT, ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT, ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT`

Reply:
Not quite following the interpret part, could you explain?

The docstring is added in the next PR in stack, a mistake when splitting the commits but i'll just leave it as is